packages/wrangler/src/config/validation.ts

```diff
@@ -461,6 +461,84 @@ function normalizeAndValidateModulePaths(
   return mapping;
 }
 
+/**
+ * Check whether a value has the shape of a route, which can be a string
+ * or an object that looks like {pattern: string, zone_id: string }
+ */
+function isValidRouteValue(item: unknown): boolean {
+  return (
+    !!item &&
+    (typeof item === "string" ||
+      (typeof item === "object" &&
+        // @ts-expect-error how do i refine this type?
+        typeof item.pattern === "string" &&
+        // @ts-expect-error how do i refine this type?
+        typeof item.zone_id === "string" &&
+        Object.keys(item).length === 2))
+  );
+}
```

Comment:
Maybe something like this, using a predicate with `RoutesValue` being the type to narrow too so that the object checks should no longer necessary like `typeof item.pattern === "string"`  
```suggestion
type RoutesValue = {
pattern: string,
zone_id: string
} | string

function isValidRouteValue(item: RoutesValue): item is RoutesValue {
return item === "string" || Object.keys(item).length === 2 
}
```

Reply:
You would still want to check that the keys were pattern and zone_id.

Reply:
good shout. but I'm trying to avoid using a type since it's likely the object _won't_ match the type, and that's what we're testing. unless that's the idiomatic way to do this? I'd be surprised if so.

Reply:
```ts
/**
 * Check whether a value has the shape of a route, which can be a string
 * or an object that looks like {pattern: string, zone_id: string }
 */
function isValidRouteValue(item: unknown): boolean {
  return (
    !!item &&
    (typeof item === "string" ||
      (typeof item === "object" &&
        hasProperty("pattern", item) &&
        typeof item.pattern === "string" &&
        hasProperty("zone_id", item) &&
        typeof item.zone_id === "string" &&
        Object.keys(item).length === 2))
  );
}

type HasProperty<U, K extends PropertyKey> = K extends keyof U
  ? U
  : U & Record<K, unknown>;
function hasProperty<K extends PropertyKey, T>(
  k: K,
  o: T
): o is HasProperty<T, K> {
  return k in o;
}
```

Reply:
😳

Reply:
Actually I think you can drop the `HasProperty` type and just use:

```ts
function hasProperty<K extends PropertyKey, T>(
  k: K,
  o: T
): o is T & Record<K, unknown> {
  return k in o;
}
```

Reply:
I thought a little more about it and narrowed the type and validated the object but kept the predicate. Should be now doing what you were doing. 
```typescript
type RoutesValue =
  | string
  | {
      pattern: string;
      zone_id: string;
    };

function isValidRouteValue(item: unknown): item is RoutesValue {
  if (typeof item === "string") {
    return true;
  }
  if (typeof item === "object" && item !== null) {
    return (
      Object.keys(item as RoutesValue).length === 2 &&
      "pattern" in item &&
      "zone_id" in item
    );
  }
  return false;
}
```

Reply:
> Actually I think you can drop the `HasProperty` type and just use:
> 
> ```ts
> function hasProperty<K extends PropertyKey, T>(
>   k: K,
>   o: T
> ): o is T & Record<K, unknown> {
>   return k in o;
> }
> ```

I like this! However, I still would suggest using the predicate function to narrow the type to the shape we expect so it's not just a boolean response.

Reply:
But we don't want to narrow it to what we expect, indeed we start with the expectation that it _isn't_ matching the type, hence the runtime checks.